File: examples/public-api/javascript/01-basic-workflow.ts
/**
 * Example 01: Basic Workflow Execution
 *
 * This example demonstrates the most common use case:
 * 1. Execute a workflow with inputs
 * 2. Wait for completion using polling
 * 3. Retrieve the results
 *
 * Run: npx tsx 01-basic-workflow.ts
 */

import "dotenv/config";
import { FlowMaestroClient } from "@flowmaestro/sdk";

// Configuration from environment
const API_KEY = process.env.FLOWMAESTRO_API_KEY!;
const BASE_URL = process.env.FLOWMAESTRO_BASE_URL;
const WORKFLOW_ID = process.env.WORKFLOW_ID!;

async function main() {
    // Validate configuration
    if (!API_KEY || API_KEY === "fm_live_your_api_key_here") {
        console.error("Error: Please set FLOWMAESTRO_API_KEY in your .env file");
        process.exit(1);
    }
    if (!WORKFLOW_ID || WORKFLOW_ID === "wf_your_workflow_id") {
        console.error("Error: Please set WORKFLOW_ID in your .env file");
        process.exit(1);
    }

    // Initialize the client
    const client = new FlowMaestroClient({
        apiKey: API_KEY,
        baseUrl: BASE_URL
    });

    console.log("FlowMaestro Basic Workflow Execution Example\n");
    console.log("=".repeat(50));

    try {
        // Step 1: Get workflow details to understand required inputs
        console.log("\n1. Fetching workflow details...");
        const { data: workflow } = await client.workflows.get(WORKFLOW_ID);
        console.log(`   Workflow: ${workflow.name}`);
        console.log(`   Description: ${workflow.description || "N/A"}`);

        if (workflow.inputs) {
            console.log("   Required inputs:");
            for (const [key, input] of Object.entries(workflow.inputs)) {
                console.log(`     - ${key}: ${input.type}${input.required ? " (required)" : ""}`);
            }
        }

        // Step 2: Execute the workflow
        console.log("\n2. Executing workflow...");
        const { data: execution } = await client.workflows.execute(WORKFLOW_ID, {
            inputs: {
                // Replace these with actual inputs for your workflow
                name: "John Doe",
                email: "john@example.com"
            }
        });
        console.log(`   Execution ID: ${execution.execution_id}`);
        console.log(`   Status: ${execution.status}`);

        // Step 3: Wait for completion using polling
        console.log("\n3. Waiting for completion...");
        const startTime = Date.now();

        const result = await client.executions.waitForCompletion(execution.execution_id, {
            pollInterval: 1000, // Poll every 1 second
            timeout: 120000 // Timeout after 2 minutes
        });

        const duration = ((Date.now() - startTime) / 1000).toFixed(2);
        console.log(`   Completed in ${duration}s`);

        // Step 4: Display results
        console.log("\n4. Results:");
        console.log(`   Status: ${result.status}`);

        if (result.status === "completed") {
            console.log("   Outputs:");
            if (result.outputs) {
                for (const [key, value] of Object.entries(result.outputs)) {
                    const displayValue =
                        typeof value === "object" ? JSON.stringify(value, null, 2) : value;
                    console.log(`     ${key}: ${displayValue}`);
                }
            } else {
                console.log("     (no outputs)");
            }
        } else if (result.status === "failed") {
            console.log(`   Error: ${result.error}`);
        }

        console.log("\n" + "=".repeat(50));
        console.log("Example completed successfully!");
    } catch (error) {
        console.error("\nError:", error instanceof Error ? error.message : error);
        process.exit(1);
    }
}

main();

File: examples/public-api/python/01_basic_workflow.py
"""
Example 01: Basic Workflow Execution

This example demonstrates the most common use case:
1. Execute a workflow with inputs
2. Wait for completion using polling
3. Retrieve the results

Run: python 01_basic_workflow.py
"""

import os
import sys
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
API_KEY = os.getenv("FLOWMAESTRO_API_KEY")
BASE_URL = os.getenv("FLOWMAESTRO_BASE_URL")
WORKFLOW_ID = os.getenv("WORKFLOW_ID")


def main():
    # Validate configuration
    if not API_KEY or API_KEY == "fm_live_your_api_key_here":
        print("Error: Please set FLOWMAESTRO_API_KEY in your .env file")
        sys.exit(1)
    if not WORKFLOW_ID or WORKFLOW_ID == "wf_your_workflow_id":
        print("Error: Please set WORKFLOW_ID in your .env file")
        sys.exit(1)

    # Import SDK (after validation to give better error messages)
    from flowmaestro import FlowMaestroClient

    # Initialize the client
    client = FlowMaestroClient(
        api_key=API_KEY,
        base_url=BASE_URL if BASE_URL else None
    )

    print("FlowMaestro Basic Workflow Execution Example\n")
    print("=" * 50)

    try:
        # Step 1: Get workflow details
        print("\n1. Fetching workflow details...")
        response = client.workflows.get(WORKFLOW_ID)
        workflow = response["data"]
        print(f"   Workflow: {workflow['name']}")
        print(f"   Description: {workflow.get('description', 'N/A')}")

        if workflow.get("inputs"):
            print("   Required inputs:")
            for key, input_def in workflow["inputs"].items():
                required = " (required)" if input_def.get("required") else ""
                print(f"     - {key}: {input_def['type']}{required}")

        # Step 2: Execute the workflow
        print("\n2. Executing workflow...")
        response = client.workflows.execute(
            WORKFLOW_ID,
            inputs={
                # Replace these with actual inputs for your workflow
                "name": "John Doe",
                "email": "john@example.com"
            }
        )
        execution = response["data"]
        print(f"   Execution ID: {execution['execution_id']}")
        print(f"   Status: {execution['status']}")

        # Step 3: Wait for completion
        print("\n3. Waiting for completion...")
        import time
        start_time = time.time()

        result = client.executions.wait_for_completion(
            execution["execution_id"],
            poll_interval=1.0,  # Poll every 1 second
            timeout=120.0  # Timeout after 2 minutes
        )

        duration = time.time() - start_time
        print(f"   Completed in {duration:.2f}s")

        # Step 4: Display results
        print("\n4. Results:")
        print(f"   Status: {result['status']}")

        if result["status"] == "completed":
            print("   Outputs:")
            outputs = result.get("outputs", {})
            if outputs:
                for key, value in outputs.items():
                    print(f"     {key}: {value}")
            else:
                print("     (no outputs)")
        elif result["status"] == "failed":
            print(f"   Error: {result.get('error', 'Unknown error')}")

        print("\n" + "=" * 50)
        print("Example completed successfully!")

    except Exception as e:
        print(f"\nError: {e}")
        sys.exit(1)
    finally:
        client.close()


if __name__ == "__main__":
    main()
